File: src/main/java/com/leetcode/P198.java
package com.leetcode;

// https://github.com/antop-dev/algorithm/issues/400
public class P198 {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        nums[1] = Math.max(nums[1], nums[0]);
        for (int i = 2; i < nums.length; i++) {
            nums[i] = Math.max(nums[i] + nums[i - 2], nums[i - 1]);
        }
        return nums[nums.length - 1];
    }
}

File: src/main/kotlin/com/leetcode/P198.kt
package com.leetcode

// https://github.com/antop-dev/algorithm/issues/400
class P198 {
    fun rob(nums: IntArray): Int {
        val dp = IntArray(nums.size + 1)
        dp[1] = nums[0]

        for (i in 2 until dp.size) {
            // 털지 않는 경우(dp[i - 1]) or 터는 경우(dp[i - 2] + nums[i - 1])
            dp[i] = maxOf(dp[i - 1], dp[i - 2] + nums[i - 1])
        }

        return dp.last()
    }
}
